### PR TITLE
PMM-9672: new routing system

### PR DESCRIPTION
**Pull request by fabio-silva**
**What this PR does / why we need it**: Uses same routing system as Grafana, making it consistent and easier to have breadcrumbs and tabbed pages.

**Which issue(s) this PR fixes**: https://jira.percona.com/browse/PMM-9672

*Details*

- Use `<Page navModel={navModel}><Page.Contents>{children}</Page.Contents></Page>` to render a page;
- `const navModel = usePerconaNavModel('dbclusters');`
- Example of page model:

```
  export const PMM_DBAAS_PAGE: NavModelItem = {
      id: 'dbaas',
      text: 'DBaaS',
      subTitle: 'Percona DBaaS',
      icon: 'database',
      url: `${getConfig().appSubUrl}/dbaas`,
      breadcrumbs: [
        {
          title: 'DBaaS',
          url: `${getConfig().appSubUrl}/dbaas`,
        },
      ],
      children: [
        {
          id: 'kubernetes',
          text: 'Kubernetes Cluster',
          url: `${getConfig().appSubUrl}/dbaas/kubernetes`,
          hideFromMenu: true,
        },
        {
          id: 'dbclusters',
          text: 'DB Cluster',
          url: `${getConfig().appSubUrl}/dbaas/dbclusters`,
          hideFromMenu: true,
        },
      ],
    };
```
- Dispatch action on navbar to update store with model: `dispatch(updateNavIndex(PMM_DBAAS_PAGE));`

Screenshot 2022-03-25 at 10 16 53
Screenshot 2022-03-25 at 09 51 00
Screenshot 2022-03-25 at 10 16 40
Screenshot 2022-03-25 at 10 16 47

### CODE REVIEW

File: public/app/core/components/NavBar/constants.ts [hunk 1]
```diff
@@ -0,0 +1,219 @@
+import { NavModelItem } from '@grafana/data';
+import { getConfig } from 'app/core/config';
```
Comment (tiagomotasantos) [inline public/app/core/components/NavBar/constants.ts:2]: Can't we import config directly? Instead of getConfig.

File: public/app/core/hooks/useNavModel.ts [hunk 2]
```diff
@@ -5,5 +5,7 @@ import { NavModel } from '@grafana/data';
 
 export const useNavModel = (id: string): NavModel => {
   const navIndex = useSelector((state: StoreState) => state.navIndex);
-  return getNavModel(navIndex, id);
+  const model = getNavModel(navIndex, id);
```
Comment (tiagomotasantos) [inline public/app/core/hooks/useNavModel.ts:8]: Why this change?

**Review by tiagomotasantos — COMMENTED** [review]
Very well done with the routing change 👏

Also I prefer this new design for the settings page but I'm wondering if this it the right PR to do it. Since we decided to re design the settings page it will be strange from the users perspective to see a change in the design and then when we work on a final version see another change.

File: public/app/percona/inventory/Tabs/Nodes.tsx [hunk 3]
```diff
@@ -79,81 +83,95 @@ export const NodesTab = () => {
...
+                            />
+                          }
+                        />
+                        <HorizontalGroup justify="space-between" spacing="md">
+                          <Button variant="secondary" size="md" onClick={() => setModalVisible(false)}>
+                            Cancel
+                          </Button>
+                          <Button
+                            size="md"
+                            onClick={() => {
```
Comment (TomasBlazek3PG) [inline public/app/percona/inventory/Tabs/Nodes.tsx:143]: Wouldn't it be better to have this as a callback?

File: public/app/percona/settings/components/SSHKey/SSHKey.tsx [hunk 4]
```diff
@@ -1,53 +1,92 @@
...
   const [loading, setLoading] = useState(false);
+  const [generateToken] = useCancelToken();
+  const { result: settings } = useSelector(getPerconaSettings);
+  const dispatch = useAppDispatch();
+  const navModel = usePerconaNavModel('settings-ssh');
+  const { sshKey } = settings!;
   const isEqual = (a: string, b: string) => !(a && !b) || a === b;
-  const applyChanges = ({ key }: { key: string }) => updateSettings({ ssh_key: key }, setLoading);
+
+  const applyChanges = async ({ key }: { key: string }) => {
```
Comment (TomasBlazek3PG) [inline public/app/percona/settings/components/SSHKey/SSHKey.tsx:35]: Shouldn't this be useCallback?

File: public/app/percona/shared/components/hooks/catchCancellationError.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import { SerializedError } from '@reduxjs/toolkit';
+import { PERCONA_CANCELLED_ERROR_NAME } from '../../core/constants';
+import { isApiCancelError } from '../../helpers/api';
+
+// Used to safely return from a side effect that might trigger when the component unmounted
+// and some promise resolved in the meanwhile
+export const useCatchCancellationError = () => {
+  const catchFromAsyncThunkAction = async <T>(p: Promise<T>): Promise<T | void> => {
+    const data = await p;
+    //@ts-ignore
```
Comment (TomasBlazek3PG) [inline public/app/percona/shared/components/hooks/catchCancellationError.ts:11]: Why is it here?

**Review by tiagomotasantos — APPROVED** [review]

File: public/app/percona/shared/components/hooks/catchCancellationError.ts (hunk 5 above)
Comment (fabio-silva) [reply public/app/percona/shared/components/hooks/catchCancellationError.ts]: Yeah, typescript wouldn't recognize the error since I'm using a generic type. Just used `&` to allow that

File: public/app/percona/inventory/Tabs/Nodes.tsx (hunk 3 above)
Comment (fabio-silva) [reply public/app/percona/inventory/Tabs/Nodes.tsx]: Indeed 👍

File: public/app/percona/settings/components/SSHKey/SSHKey.tsx (hunk 4 above)
Comment (fabio-silva) [reply public/app/percona/settings/components/SSHKey/SSHKey.tsx]: Yep 👍

File: public/app/routes/routes.tsx [hunk 6]
```diff
@@ -374,75 +374,216 @@ export function getAppRoutes(): RouteDescriptor[] {
...
         () =>
           import(
-            /* webpackChunkName: "IntegratedAlertingPage" */ 'app/percona/integrated-alerting/IntegratedAlertingPage'
+            /* webpackChunkName: "IntegratedAlertingChannels" */ 'app/percona/integrated-alerting/components/NotificationChannel/NotificationChannel'
           )
       ),
     },
     {
-      path: '/integrated-alerting/:tab',
+      path: '/integrated-alerting/alert-rules',
```
Comment (solovevayaroslavna) [inline public/app/routes/routes.tsx:408]: like)

File: public/app/routes/routes.tsx [hunk 7]
```diff
@@ -374,75 +374,216 @@ export function getAppRoutes(): RouteDescriptor[] {
...
     {
       path: '/backup',
-      component: SafeDynamicImport(() => import(/* webpackChunkName: "BackupPage" */ 'app/percona/backup/BackupPage')),
+      // eslint-disable-next-line react/display-name
+      component: () => <Redirect to="/backup/inventory" />,
     },
     {
-      path: '/backup/:tab',
-      component: SafeDynamicImport(() => import(/* webpackChunkName: "BackupPage" */ 'app/percona/backup/BackupPage')),
+      path: '/backup/inventory',
```
Comment (solovevayaroslavna) [inline public/app/routes/routes.tsx:442]: It's become much more readable for me now =))))

**Review by solovevayaroslavna — CHANGES_REQUESTED** [review]
A lot of work has been done! Thanks! I would like to draw attention to the fact that we need to stick to the same solutions in the tests, this will improve readability. So I would ask to fix some things on toHaveLength and toHaveTextContent.

File: packages/grafana-ui/src/components/Tabs/TabsBar.tsx [hunk 8]
```diff
@@ -22,19 +24,24 @@ const getTabsBarStyles = stylesFactory((theme: GrafanaTheme2, hideBorder = false
...
       height: 41px;
+      flex-direction: ${!!vertical ? 'column' : 'row'};
     `,
   };
 });
 
-export const TabsBar = React.forwardRef<HTMLDivElement, Props>(({ children, className, hideBorder }, ref) => {
-  const theme = useTheme2();
-  const tabsStyles = getTabsBarStyles(theme, hideBorder);
+export const TabsBar = React.forwardRef<HTMLDivElement, Props>(
```
[the remark this answers is not in the document]
Comment (fabio-silva) [reply packages/grafana-ui/src/components/Tabs/TabsBar.tsx:32]: This component is part of grafana's code, "ui" package, which becomes published as `@grafana/ui` on npm, that's why shouldn't move it :)

File: pkg/api/index.go [hunk 9]
```diff
@@ -171,10 +171,6 @@ func (hs *HTTPServer) getNavTree(c *models.ReqContext, hasEditPerm bool) ([]*dto
 	}
 
 	dashboardChildLinks := hs.buildDashboardNavLinks(c, hasEditPerm)
-	inventoryChildNavs := []*dtos.NavLink{
```
[the remark this answers is not in the document]
Comment (fabio-silva) [reply pkg/api/index.go:174]: Indeed. Before, we declared some routes on the backend. But because this route might not be available for everyone, all Percona's routes are defined on the Navbar, on the `buildInventoryAndSettings` function.

File: public/app/core/components/PageHeader/PageHeader.tsx [hunk 10]
```diff
@@ -124,23 +141,43 @@ function renderTitle(title: string, breadcrumbs: NavModelBreadcrumb[]) {
...
 
   return <h1 className="page-header__title">{breadcrumbsResult}</h1>;
 }
 
-const getStyles = (theme: GrafanaTheme2) => ({
-  headerCanvas: css`
-    background: ${theme.colors.background.canvas};
-  `,
-});
+const getStyles = (theme: GrafanaTheme2) => {
```
[the remark this answers is not in the document]
Comment (fabio-silva) [reply public/app/core/components/PageHeader/PageHeader.tsx:167]: Grafana's code :)

File: public/app/core/reducers/navModel.ts [hunk 11]
```diff
@@ -63,10 +63,16 @@ export const navIndexReducer = (state: NavIndex = initialState, action: AnyActio
     const newPages: NavIndex = {};
     const payload = action.payload;
 
-    for (const node of payload.children!) {
-      newPages[node.id!] = {
-        ...node,
-        parentItem: payload,
+    if (payload.children && payload.children.length) {
+      for (const node of payload.children) {
+        newPages[node.id!] = {
```
[the remark this answers is not in the document]
Comment (fabio-silva) [reply public/app/core/reducers/navModel.ts]: Well, the `!` at the end tells Typescript that this `id` won't be undefined at this point, so it won't complain about it. I followed the same approach as they do some lines above. But you're right, better not to risk it. I added a safety check now.

File: public/app/features/alerting/unified/utils/redux.ts [hunk 12]
```diff
@@ -13,15 +15,21 @@ export interface AsyncRequestState<T> {
...
-  dispatched: false,
-});
+interface AsyncInitialRequestState<T> extends AsyncRequestState<T> {
+  result: T;
+}
+
+export const createInitialAsyncRequestState = <T>(
+  state: T
+): Pick<AsyncInitialRequestState<T>, 'loading' | 'dispatched' | 'result' | 'error'> =>
+  Object.freeze({
```
[the remark this answers is not in the document]
Comment (fabio-silva) [reply public/app/features/alerting/unified/utils/redux.ts:25]: This code was already in place, I changed it a bit to allow an initial state, instead of `undefined`. But yes, by doing a freeze, they make sure this initial state is not altered directly. When working with reducers, it's a rule to always create a new state from the previous one rather then changing the current one. `Object.freeze` makes sure that this particular object won't change.

File: public/app/percona/add-instance/components/AddInstance/AddInstance.test.tsx [hunk 13]
```diff
@@ -7,18 +7,28 @@ jest.mock('app/percona/settings/Settings.service');
...
     expect(screen.getAllByRole('button').length).toEqual(instanceList.length);
     instanceList.forEach((item) => {
       expect(screen.getByTestId(`${item.type}-instance`)).toBeInTheDocument();
     });
   });
 
+  it('should render azure option', async () => {
+    await waitFor(() => render(<AddInstance showAzure onSelectInstanceType={() => {}} />));
+
+    expect(screen.getAllByRole('button').length).toEqual(instanceList.length + 1);
```
[the remark this answers is not in the document]
Comment (fabio-silva) [reply public/app/percona/add-instance/components/AddInstance/AddInstance.test.tsx]: 👍

File: public/app/percona/backup/components/ScheduledBackups/ScheduledBackups.test.tsx [hunk 14]
```diff
@@ -1,17 +1,27 @@
...
+          percona: {
+            user: { isAuthorized: true, isPlatformUser: false },
+            settings: { result: { backupEnabled: true, isConnectedToPortal: false } },
+          },
+        } as StoreState)}
+      >
+        <ScheduledBackups />
+      </Provider>
+    );
+    await screen.findByText('Backup 1');
```
[the remark this answers is not in the document]
Comment (fabio-silva) [reply public/app/percona/backup/components/ScheduledBackups/ScheduledBackups.test.tsx:24]: Indeed. I usually avoid `waitFor`, because it doesn't tell us much. Waiting for `Backup 1` means the data was loaded already on the table, which is what visually tells us the page is done. It's a more "UI" approach of testing, closer to what users see

File: public/app/percona/check/components/AllChecksTab/AllChecksTab.test.tsx [hunk 15]
```diff
@@ -78,23 +105,35 @@ describe('AllChecksTab::', () => {
...
-    expect(tbody.querySelectorAll('tr > td')[5]).toHaveTextContent('Test disabled');
-    expect(tbody.querySelectorAll('tr > td')[6]).toHaveTextContent('test disabled description');
-    expect(tbody.querySelectorAll('tr > td')[7]).toHaveTextContent(Messages.disabled);
-    expect(tbody.querySelectorAll('tr > td')[8]).toHaveTextContent(Interval.RARE);
-    expect(tbody.querySelectorAll('tr > td')[9]).toHaveTextContent(Messages.enable);
+    expect(screen.getAllByTestId('db-checks-all-checks-table')).toHaveLength(1);
+    expect(screen.getAllByTestId('db-checks-all-checks-thead')).toHaveLength(1);
+    expect(screen.getAllByTestId('db-checks-all-checks-tbody')).toHaveLength(1);
+    expect(cells).toHaveLength(10);
+    expect(cells[0].textContent).toBe('Test');
```
[the remark this answers is not in the document]
Comment (fabio-silva) [reply public/app/percona/check/components/AllChecksTab/AllChecksTab.test.tsx]: 👍

File: public/app/percona/check/components/ServiceChecks/ServiceChecks.test.tsx [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+      >
+        <ServiceChecks
+          {...getRouteComponentProps({
+            match: { params: { service: '/service_1/' }, isExact: true, path: '', url: '' },
+          })}
+        />
+      </Provider>
+    );
+    await waitForElementToBeRemoved(() => screen.getByTestId('table-loading'));
+    expect(screen.getByTestId('page-service').textContent).toBe('Failed Checks for service "Service One"');
```
[the remark this answers is not in the document]
Comment (fabio-silva) [reply public/app/percona/check/components/ServiceChecks/ServiceChecks.test.tsx]: 👍

File: public/app/percona/integrated-alerting/components/AlertRules/AlertRules.test.tsx [hunk 17]
```diff
@@ -44,47 +47,85 @@ describe('AlertRules', () => {
...
+          },
+        } as StoreState)}
+      >
+        <AlertRules />
+      </Provider>
+    );
+
+    await waitForElementToBeRemoved(() => screen.getByTestId('table-loading'));
+
+    expect(screen.getAllByRole('row')).toHaveLength(1 + 6);
```
[the remark this answers is not in the document]
Comment (fabio-silva) [reply public/app/percona/integrated-alerting/components/AlertRules/AlertRules.test.tsx:105]: Yes! Because the first row is the table's header, so I think it's easier to follow like this

File: public/app/routes/routes.tsx (hunk 6 above)
Comment (fabio-silva) [reply public/app/routes/routes.tsx:408]: 🤘

File: public/app/routes/routes.tsx (hunk 7 above)
Comment (fabio-silva) [reply public/app/routes/routes.tsx:442]: Yes, more explicit now )

**Review by solovevayaroslavna — DISMISSED** [review]

**Review by solovevayaroslavna — APPROVED** [review]

**Review by solovevayaroslavna — APPROVED** [review]
